Hide drop indicator where block isn't allowed to drop.

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes #24174.

If a block isn't allowed to drop in a certain area, the drop indicator shouldn't show.

In addition, the draggable chip changes color and icon when hovering over an area where there are no valid drop targets.


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

1. In the post or template editor, add a few block containers of different types (Group, Gallery, Navigation, List for example) and also add a few loose blocks of different types such as Paragraph and Image.
2. Try dragging single blocks and also multiple selections of different blocks into the different containers.
3. Check that drop indicator only shows if all selected blocks being dragged are allowed in the container.
4. Try dragging single and multiple blocks around at root level and check this always works well.
5. Try dragging blocks that should only exist _inside_ certain containers (List Item, Navigation Link for example) outside of their containers - no drop indicators should show.

~~Note the transparency of the draggable chip isn't yet quite in sync with the display of valid drop targets. This can be fine-tuned but it would be great to get some initial feedback on the general approach first!~~
Update: changes to draggable chip should now be in sync with showing the drop indicator.

## Screenshots or screencast <!-- if applicable -->

https://github.com/WordPress/gutenberg/assets/8096000/5add2ee7-3c32-463b-a5fc-eaea21114154

https://github.com/WordPress/gutenberg/assets/8096000/661e1475-d880-4225-a898-bfd2185d4dc3

tellthemachines committed:
Hide drop indicator where block isn't allowed to drop.

tellthemachines committed:
Don't show indicator when drop outside of allowed parent

tellthemachines committed:
logic to change the draggable icon

tellthemachines committed:
Make chip transparent when dragging over undroppable area.

tellthemachines committed:
Remove unused prop

andrewserong commented on packages/block-editor/src/components/use-block-drop-zone/index.js:
```diff
@@ -191,6 +194,65 @@ export function getDropTargetPosition(
+ * @param {string}   targetClientId        The client id of the target.
+ * @return {boolean} Whether the dragged blocks can be dropped on the target.
+ */
+export function useIsDropTargetValid( draggedBlockClientIds, targetClientId ) {
```
This hook appears to get called even when a drag event isn't happening. Is it possible to move some of the logic to the throttled `useCallback` so that we only figure out the allowed blocks in the throttled function that gets called while a user is dragging? Ideally it'd be good to limit what gets called before the drag actions occur, but it might not be possible to move everything over 🤔

tellthemachines replied:
This could be refactored to a simple function taking `getBlockType, allowedBlocks, draggedBlockNames, targetBlockName` as parameters and then called inside the throttle.  I'm unsure if determining `isBlockDroppingAllowed` inside the throttle will cause any lag but let's give it a try 😄

andrewserong commented on packages/block-editor/src/components/block-draggable/index.js:
```diff
@@ -102,7 +161,12 @@ const BlockDraggable = ( {
+					count={ clientIds.length }
+					icon={ icon }
+					clientIds={ clientIds }
+					targetClientId={ targetClientId }
```
Is this addition currently doing anything? It looks like `BlockDraggableChip` isn't currently using it?

tellthemachines replied:
yeah I must have left that in while experimenting 😄

andrewserong commented on packages/block-editor/src/components/block-draggable/index.js:
```diff
@@ -20,31 +22,39 @@ const BlockDraggable = ( {
+
+	const [ targetClientId, setTargetClientId ] = useState( null );
+
+	const isDropTargetValid = useIsDropTargetValid( clientIds, targetClientId );
```
Does the `BlockDraggable` component need to know whether the drop target is valid? Could we use the `getBlockInsertionPoint` state to figure out if an insertion point is currently set instead?

tellthemachines replied:
I don't think so;  given that [`getBlockInsertionPoint`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/store/selectors.js#L1404) is always meant to return an index, it would be hard to work out which targets are valid.

andrewserong replied:
Ah, gotcha! Looks like there's another selector [`isBlockInsertionPointVisible`](https://github.com/WordPress/gutenberg/blob/e07802b3d2395c4563c6e8a8013bb2224e0927d7/packages/block-editor/src/store/selectors.js#L1442) which handles the `null` case. Could that work if we'd called `hideInsertionPoint` in the main throttled `useCallback` in `useBlockDropZone`?

tellthemachines replied:
Oh nice find! Lemme try that out

tellthemachines replied:
Ok checking if the insertion point is visible does fix the discrepancy between transparent drag chip and inserter, but I found it necessary to keep the check for valid insertion point in place, because otherwise the chip keeps flickering every time the insertion point switches places. Still it's working much better now so thanks for the tip!

andrewserong reviewed (commented):
Nice work, it feels much better not showing a drop indicator when you can't drop somewhere to me, and I also don't mind the opacity to flag that a drop isn't available. It's really nice seeing some visual feedback between a valid area and a not valid area 👍

Some of my review comments might be a bit out of order, I'm sorry, so I'll try to line up some of my thinking as I was looking through the code change:

Instead of toggling disabling the drop zone altogether, would it be possible to call [`hideInsertionPoint`](https://github.com/WordPress/gutenberg/blob/813318665d6c15a3a200d0b73beab5021457171e/packages/block-editor/src/store/actions.js#L655) instead, from within the  throttled `useCallback` in `useBlockDropZone` (or update the [`showInsertionPoint` call](https://github.com/WordPress/gutenberg/blob/d4887ec26c02ea4cbd38bff99bf4b4e09917fb25/packages/block-editor/src/components/use-block-drop-zone/index.js#L295) to pass in `null`)? I haven't used that before, but from looking at the `BlockDropZonePopover` component, it looks like it checks [that state here](https://github.com/WordPress/gutenberg/blob/96e5dc737b910b38a715d50b23c26af0c593a67d/packages/block-editor/src/components/block-popover/drop-zone.js#L25), so I was wondering if that might be a simpler way to hide the drop indicator? (I'm assuming that when it's hidden in state, that results in the drop indicator disappearing altogether, but I haven't tested it)

Then, potentially `BlockDraggable` could also use `getBlockInsertionPoint` to see if it's nullish? That might also help with syncing between the two bits of code, as the opacity change would only happen when the `useCallback` fires the changes in `useBlockDropZone` 🤔

I'm not sure if all that would work, there's quite likely a bunch of complexity there that I'm not aware of, so apologies if that's not a helpful idea, or if it's something you already tried while exploring this!

tellthemachines committed:
Check insertion point visibility

andrewserong commented on packages/block-editor/src/components/block-draggable/index.js:
```diff
@@ -61,6 +78,55 @@ const BlockDraggable = ( {
+			) {
+				setTargetClientId( newTargetClientId );
+			}
+			// Update the body class to reflect if drop target is valid.
+			// This has to be done on the document body because the draggable
+			// chip is rendered outside of the editor iframe.
+			if ( ! isDropTargetValid && ! visibleInserter ) {
+				window?.document?.body?.classList?.add(
+					'block-draggable-invalid-drag-token'
+				);
+			} else {
+				window?.document?.body?.classList?.remove(
+					'block-draggable-invalid-drag-token'
+				);
+			}
+		};
```
If we wanted to avoid attaching classes to the body element, would it work to pass an `isDisabled` boolean to `BlockDraggableChip` and have that component output an extra classname alongside `block-editor-block-draggable-chip-wrapper` if that flag is set? Or, alternately, update `BlockDraggableChip` to accept a `className` param?

Or, is it better to have the classname be output on the `body` so that potentially other kinds of drag chip components could leverage it, too? 🤔 

Just thinking out-loud! 😄

tellthemachines replied:
I tried all of that already 😄 

The problem is that Draggable makes a copy of the chip in its [start function](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/draggable/index.tsx#L100C2-L100C2) and that stored copy is what gets moved around, so its state never changes from the initial render. That's why I'm using the body classname to update the value of a CSS custom property: it was the simplest way I found to affect the copy of the draggable chip in real time. I also wanted to avoid any changes to Draggable itself, because it's a generic component that shouldn't have to know anything about what it's dragging.

Fwiw Draggable adds a class to body while dragging in order to change the cursor, which is what gave me the idea in the first place.

andrewserong replied:
Ah, of course. Sounds good to me! Thanks for the explanation 🙂

ramonjd commented on packages/block-editor/src/components/block-draggable/index.js:
```diff
@@ -61,6 +78,55 @@ const BlockDraggable = ( {
+			// and if it's a container block
+			if (
+				targetClientId !== newTargetClientId &&
+				getBlockListSettings( newTargetClientId )
```
Is `getBlockListSettings( newTargetClientId )` checking for nested blocks to see if it's a container?

I think calling a select outside `useSelect` is an anti-pattern as it risks not returning fresh data when the store is updated.

I tested by updating the list settings for a block via `wp.data.dispatch('core/block-editor').updateBlockListSettings()` and it looks like the return state is updated, I guess because a rerender is triggered by another prop. So it's working in that case as far as I can tell.

cc @jsnajdr in case I missed something.

tellthemachines replied:
I don't _think_ it's an anti-pattern, we use select functions outside of `useSelect` all over the place 😅 

`getBlockListSettings` was the most reliable way I found of checking whether a block is a container, given that only containers are block lists (and they might not have any inner blocks if they're empty). I'm definitely open to ideas on this one though 😄

tellthemachines replied:
Actually scratch that, I don't think we need to check for the container anymore 😂 now we're using the inserter visibility as per @andrewserong's suggestion. 

Still keen to hear about preferred code patterns though 👀

jsnajdr replied:
Here `getBlockListSettings` is called in an event handler (`onDragOver`) and that's fine. We want to read the fresh value when the event is being dispatched, and that's what happens here. No rerenders need to be triggered when the value changes.

The only thing I would change here is to write the code, for clarity, as two separate `useSelect` calls:
```js
const { getBlockListSettings } = useSelect( blockEditorStore );
// or alternatively
const { getBlockListSettings } = useRegistry().select( blockEditorStore );

const { ...rest } = useSelect( ( select ) => ... );
```
One of them just gets a reference to the selector, without any store subscriptions or reactivity, and the second reads data in the usual reactive way.

ramonjd replied:
Thanks for confirming and setting me straight on that one, @jsnajdr! 

And sorry about the misdirection @tellthemachines

tellthemachines replied:
Ok thanks for the clarification!

tellthemachines replied:
> The only thing I would change here is to write the code, for clarity, as two separate useSelect calls

There are other selectors we're grabbing directly and using inside the effect such as getBlockType and getAllowedBlocks. I guess following that logic they should all be in a separate useSelect call?

ramonjd commented:
This a sweet improvement. 🎉 

I can drop things anywhere in trunk!

https://github.com/WordPress/gutenberg/assets/6458278/c7c58361-26e7-4105-b795-b87fe3ba6568


This PR has more sensible feedback:
https://github.com/WordPress/gutenberg/assets/6458278/311a1276-3b84-452a-b198-395a6a3657a1

I just pinged about a possible risk of calling the `getBlockListSettings` selector outside of `useSelect` - might be nothing.

tellthemachines committed:
Move some code around

tellthemachines commented:
Thanks for the feedback and testing folks! I think all the code suggestions have been addressed so far. The experience is much improved but I'm now seeing a slight flakiness when moving over large non-container blocks such as Image. I think that might be due to the removal of the check for container status with `getBlockListSettings`. I removed it because it was outputting a false negative for the root container, but might have to think of another way of not flagging discrete blocks as "undroppable" 🤔

andrewserong commented on packages/block-editor/src/components/block-draggable/draggable-chip.js:
```diff
@@ -13,7 +13,10 @@ import BlockIcon from '../block-icon';
-		<div className="block-editor-block-draggable-chip-wrapper">
+		<div
+			className="block-editor-block-draggable-chip-wrapper"
+			style={ { opacity: 'var(--wp--block-draggable-valid-opacity)' } }
```
Why is this applied via an inline style and CSS variable instead of by classname and direct value in `style.scss`? Or to put it slightly differently, does the JS version of the component need to have an opinion about how it's styled?

tellthemachines replied:
Good question! I don't think it has to be applied as an inline style; I initially added it that way for the sake of development speed. It _does_ have to be a CSS custom property, because its value needs to change when the body class gets added, and given that the chip we see on the page is [a stored copy of itself](https://github.com/WordPress/gutenberg/pull/56843#discussion_r1424849732) changing classes on the chip itself won't work. Essentially the custom property is acting like a ref to the element, so we can manipulate its value from outside.

andrewserong replied:
> It does have to be a CSS custom property, because its value needs to change when the body class gets added

Oh, I was imagining something like the following:

```
.block-editor-block-draggable-chip {
	opacity: 1;
}

.block-draggable-invalid-drag-token {
	.block-editor-block-draggable-chip {
		opacity: 0.6;
	}
}
```

Or does that not work? (Totally fine to go with a CSS variable by the way, just trying to understand what the limitations are 🙂)

tellthemachines replied:
Oh right, yeah that should work too! Would that be preferable to a custom property? I don't think specificity matters either way.

andrewserong replied:
I don't feel strongly either way, but I'd lean slightly more toward the nested selectors so that it's easy to swap out for (or combine) other styles like background color, border, or other subtle styling changes that folks might want to explore in the future.

tellthemachines replied:
Oh good point, yeah that would make it easier to change styles altogether if we wanted to.

andrewserong commented on packages/block-editor/src/components/block-draggable/index.js:
```diff
@@ -61,6 +86,64 @@ const BlockDraggable = ( {
+			return;
+		}
+
+		const onDragOver = ( event ) => {
```
I think this callback might need to be throttled similar to the code in `useBlockDropZone` as this currently fires very frequently:

https://github.com/WordPress/gutenberg/assets/14988353/0ece29d6-b06e-4555-8ad7-c0ecd1ca76ce

tellthemachines replied:
Hrm yeah I'm wondering how best to do that given the inability to use hooks inside the effect callback 🤔

apeatling replied:
I think I ran into this issue when testing, it seemed to crash in Safari and I couldn't move the chip any more.

tellthemachines replied:
Oo wait I see there's a `throttle` util in compose that we may be able to leverage, will try that!

tellthemachines replied:
Ok the throttle seems to make it better! I set it to the same as [draggable](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/draggable/index.tsx#L208).

andrewserong replied:
That's a little better, but it's still lagging a bit for me with the throttle at `16` (it's more visible when using CPU throttling in dev tools). I think a couple of things might be at play:

* Since the `useEffect` doesn't use a dependency array, it seems that it's being called quite frequently, and potentially adding (and removing) event listeners too frequently? For example, if I increase the throttle time to `2000`, it still appears to be firing rapidly
* In `draggable`, `16ms` is needed because it's updating the position of the drag chip. Here, I'm wondering if we can use a bigger delay / throttle, if that could smooth over the flicker when dragging between image blocks in a gallery block? I.e. if we had it at `200ms` to match `useBlockDropZone` then if folks are dragging quickly between blocks, hopefully it shouldn't flicker as much?

Would it work to maybe move defining the `onDragOver` callback to be outside of the `useEffect` (possibly in a `useCallback`) so that the `useEffect` could have a dependency array and only attach the event listener once? Apologies if you've already tried that!

tellthemachines replied:
So I had a think about this 😄 and realised that we actually don't need to use local state for `targetClientId`, and we can use a dependency array for the effect, which improves things a little, but it still has to re-render whenever either `draggedBlockNames` or `visibleInserter` change. Moving the callback out of the effect makes no difference, because it calls whichever version of the callback was attached to the event listener when the effect ran, so if we want the values inside the callback to change the effect has to re-run.

I also tried increasing the throttle time and it doesn't seem to make things any worse, so might leave it at 200 or so. I just pushed an update, let me know if it improves things at all!

andrewserong replied:
Nice work! That appears to have fixed up the performance issues for me. Subjectively, without CPU throttling, it's feeling smooth now, and with CPU throttling, it seems much the same as `trunk` to me now 🎉

andrewserong commented on packages/block-editor/src/components/block-draggable/index.js:
```diff
@@ -61,6 +86,64 @@ const BlockDraggable = ( {
+	// Add a dragover event listener to the editor root to track the blocks being dragged over.
+	// The listener has to be inside the editor iframe otherwise the target isn't accessible.
+	// Check if the dragged blocks are allowed inside the target. If not, grey out the draggable.
+	useEffect( () => {
```
Is this `useEffect` and its `onDragOver` still needed? Since the `useBlockDropZone` is hiding the insertion point, could we replace the whole useEffect with something like:

```
	useEffect( () => {
		if ( ! visibleInserter ) {
			window?.document?.body?.classList?.add(
				'block-draggable-invalid-drag-token'
			);
		} else {
			window?.document?.body?.classList?.remove(
				'block-draggable-invalid-drag-token'
			);
		}
	}, [ visibleInserter ] );
```

Then the logic of when to show and hide the insertion point could live in one place, so that the `BlockDraggable` doesn't have to also perform checks with its own drag event handler.

Or is there another reason why this component needs to perform its own separate checks?

tellthemachines replied:
Yeah, so the crucial bit here is finding the `targetClientId`, which tells us the container that the draggable is currently hovering over. To find that we need a `dragover` event listener attached to the inside of the editor iframe (if it's on the outside, it won't be able to access the elements inside the iframe). We need the `targetClientId` to check if the drop target is valid, which is necessary because if the chip is hovering over a valid container but there's no drop indicator currently showing (if we're moving between drop areas for instance), we don't want it to grey out.

Unfortunately there's not much that can be done to restrict how many times this runs without slowing down the responsiveness of the drag chip to what it's hovering over. The problem here is that, unlike with the drop indicators that exist statically between all blocks so we know where they're placed from the start, we need to find out what the draggable is being dragged over, and afaik the only way to do that is through the event target of a `dragover` listener.

andrewserong replied:
> We need the targetClientId to check if the drop target is valid, which is necessary because if the chip is hovering over a valid container but there's no drop indicator currently showing (if we're moving between drop areas for instance), we don't want it to grey out.

Ah, gotcha! Thanks for the explanation, that makes sense now 👍

> Unfortunately there's not much that can be done to restrict how many times this runs without slowing down the responsiveness of the drag chip to what it's hovering over.

Yeah, it can be really tricky to get the right balance when slowing things down so that it's slow enough for performance reasons, but fast enough to feel responsive to a user! 😅

> we need to find out what the draggable is being dragged over, and afaik the only way to do that is through the event target of a dragover listener.

That sounds about right to me, too. If it winds up being a bit of a rabbithole trying to get it working smoothly, would it be worth splitting out the `useBlockDropZone` change and landing that first? Totally cool if you want to keep it all together while figuring everything out of course!

andrewserong reviewed (commented):
This is testing pretty well for me! I did notice a bit of inconsistency with when the opacity styles are applied when dragging from the list view. Sometimes the drag chip was greyed out, and sometimes it wasn't. A subtle bit was when dragging between Image blocks in a Gallery block where it flickers a little:

https://github.com/WordPress/gutenberg/assets/14988353/c9d11419-c044-457c-ba57-dcb8438f8e19

The main issue I ran into is a performance issue while dragging over blocks — I think it's the `onDragOver` event in `BlockDraggable`, which will likely need to be throttled? (Or alternately removed, and consolidate the show/hide in `useBlockDropZone`'s callback)

https://github.com/WordPress/gutenberg/assets/14988353/10023d64-90d1-4712-9cba-2112f6254711

Another potential option, if it proves to be tricky to resolve the flickering with the opacity issue, could be to split this PR into two: one PR to implement the hiding the drop indicator line, and then a separate follow-up PR to look at the opacity change for the drag chip?

tellthemachines commented:
Thanks for the further testing and feedback!

> A subtle bit was when dragging between Image blocks in a Gallery block where it flickers a little

Yeah I _think_ this is the same issue I spotted yesterday evening and commented on [here](https://github.com/WordPress/gutenberg/pull/56843#issuecomment-1853305094).

tellthemachines committed:
Throttle the event listener

tellthemachines committed:
Add dependencies to effect

andrewserong reviewed (commented):
Looks like that last commit fixed up the performance issues I ran into! 👍

I suppose this just leaves design feedback and the sometimes flickering issues to deal with?

One other thing I noticed with the list view open (and this might be the same technical issue as the flickering between blocks) is that it feels like it kinda remembers the last visibility state when you drag a block from there. In the following screengrab, it starts out stuck in faded out mode, but if I drag a block over the canvas to get it into opaque mode, I can get it stuck in that mode again:

https://github.com/WordPress/gutenberg/assets/14988353/5294cee6-73e6-47fa-8839-bb2f6040a484

The list view also uses `[data-block]` attributes, so not sure if it's related to that at all 🤔

tellthemachines committed:
Update CSS

tellthemachines commented:
> is that it feels like it kinda remembers the last visibility state when you drag a block from there. In the following screengrab, it starts out stuck in faded out mode, but if I drag a block over the canvas to get it into opaque mode, I can get it stuck in that mode again

That makes sense because the listener is tied to the editor canvas 😅 so it won't do anything unless you happen to drag the token over the block canvas, in which case the dragover listener in this PR is activated and the body class gets added or removed, which has an effect on the list view chip too. 

What's the desired outcome? Do we want the list view chip to grey out when hovering over a non-drop area? From what I see in trunk it already doesn't show the inserter anywhere the drop can't happen. 

We could either attach a listener to the list view (we can't use the same one because of the iframe around the editor) or we could make the list view chip not grey out at all in response to the body class by passing down a custom classname to attach the styles to.

andrewserong commented:
> or we could make the list view chip not grey out at all in response to the body class by passing down a custom classname to attach the styles to.

Hrm, that's an interesting idea. I think I'd be inclined to switch off the disabled class behaviour for the list view possibly? The List View renders the [<BlockDraggable>](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/list-view/block-contents.js#L77) component on its own, so perhaps another option could be to add a param to `BlockDraggable` to switch the disabling behaviour on or off? Then the List View could have it off, but the editor canvas could have it on, potentially 🤔

tellthemachines commented:
> The List View renders the component on its own, so perhaps another option could be to add a param to `BlockDraggable` to switch the disabling behaviour on or off?

That should work too! I'll switch it off for now, we can always enable it later.

andrewserong commented:
> so its BlockDraggable will render anyway

I thought it was just the `BlockDraggable` that a user starts dragging that wins out in terms of which component gets rendered? So if the `BlockDraggable` in the list view uses a prop, and a user starts dragging from the list view, it'd get different behaviour as to if a user drags from the canvas 🤔

In any case, a prop to pass a classname down sounds fine to me, too 🙂

tellthemachines committed:
Fade chip only when dragging in editor canvas

tellthemachines commented:
> I thought it was just the `BlockDraggable` that a user starts dragging that wins out in terms of which component gets rendered?

It renders before any dragging happens from what I can tell - as soon as a block toolbar appears or soon as the list view opens. So if a block toolbar has rendered on the canvas, that instance with its event handler attached will still be hanging around when the list view item starts being dragged, and we get the weird behaviour.

andrewserong commented:
> that instance with its event handler attached will still be hanging around when the list view item starts being dragged, and we get the weird behaviour.

Gotcha. Thanks again for all the explanations here, it's super helpful!

tellthemachines commented:
> Thanks again for all the explanations here, it's super helpful!

No worries, your questions have really helped me wrap my head around this 😅 you never really understand something until you try to explain it!

apeatling reviewed (commented):
The performance improvements are working well for me, I couldn't see any slow downs or sticky chips. :+1:

I wonder if we can be more explicit about what the status of the chip is when dragging over something that's not droppable? The fade out works well, but what about also using an icon that is universally considered to mean "no" or "not possible":

<img width="638" alt="Screenshot 2023-12-14 at 9 13 30 AM" src="https://github.com/WordPress/gutenberg/assets/1464705/3f3486ec-837d-4984-8cb7-a76711607b2c">

The benefit here being that it removes a lot of ambiguity around meaning.

tellthemachines commented:
> The fade out works well, but what about also using an icon that is universally considered to mean "no" or "not possible":

Oh good idea, I'll give it a try!

richtabor commented on packages/block-editor/src/components/block-draggable/style.scss:
```diff
@@ -16,6 +16,9 @@
 	user-select: none;
 	width: max-content;
+	transition:
+		background-color 0.5s linear,
```
I suggest something much quicker, but with a tiny delay (to reduce the flashes): 

`background-color 0.1s linear 0.1s, opacity 0.1s linear 0.1s;`

richtabor replied:
While we're here, a tiny tweak to the chip's shadow makes it feel nice :) 

`box-shadow: 0 4px 8px rgba($black, 0.2);`

tellthemachines replied:
Thanks, added!

richtabor commented on packages/block-editor/src/components/block-draggable/style.scss:
```diff
@@ -45,3 +48,10 @@
+.block-draggable-invalid-drag-token {
+	.block-editor-block-draggable-chip-fade .block-editor-block-draggable-chip {
+		background-color: $gray-600;
+		opacity: 0.8;
```
I don't think opacity is making it clear enough (perhaps introducing confusion even). I suggest gray-700, without the opacity. 

![CleanShot 2023-12-15 at 17 35 20](https://github.com/WordPress/gutenberg/assets/1813435/b960d8ab-58ea-4529-b747-a81a616d10ce)

richtabor commented:
> The fade out works well, but what about also using an icon that is universally considered to mean "no" or "not possible":

I like this idea too. I'd say combined with the likes of https://github.com/WordPress/gutenberg/pull/56843#discussion_r1428554617 would communicate strongly.

tellthemachines committed:
Merge trunk and solve conflicts with #57032

tellthemachines committed:
Improve disabled chip

tellthemachines commented:
Thanks for the suggestions folks! There doesn't seem to be a forbidden icon exactly like the one @apeatling suggests in our [icon library](https://wordpress.github.io/gutenberg/?path=/story/icons-icon--library) so I've used our cancel one for now. I'm not sure it works as well but we can use it to test out the idea. How would we go about adding the forbidden one? I'm guessing a new one would have to be designed to fit in with our icon aesthetic?

apeatling commented:
Oh interesting. I found it in the WordPress design library Figma. It's called "No Swatch". I did increase the stroke width slightly.

<img width="782" alt="Screenshot 2023-12-18 at 9 09 51 AM" src="https://github.com/WordPress/gutenberg/assets/1464705/2f105bc0-0eb2-4d8a-9fba-e95a19ee558e">

tellthemachines commented:
> It's called "no-swatch". I did increase the stroke width slightly.

Oooh I see, it's not an actual icon but a styled span:
<img width="673" alt="Screenshot 2023-12-19 at 10 34 52 am" src="https://github.com/WordPress/gutenberg/assets/8096000/bc54d191-d690-42c3-a85a-a49e8210918d">
Guess we can duplicate those styles for now.

tellthemachines committed:
Update to styled span.

tellthemachines commented:
Ok updated to use a span styled like a white, slightly thicker version of the no-swatch design!

SaxonF commented:
Testing really well for me. Massive improvement over what's in trunk. From a design perspective I think we can :shipit: 

https://github.com/WordPress/gutenberg/assets/1072756/a4b2eda6-2226-4b97-8947-dd8ca86810cc

apeatling reviewed (commented):
Looks great, much better and clearer now! I do wish the performance of the crossfade was better, it feels sluggish to me. I think this is shippable, but it would be great to follow up and either get that really smooth, or remove the transition.

andrewserong reviewed (commented):
> Looks great, much better and clearer now!

Yes, visually this is looking terrific! No issues or conflicts with the list view anymore, either 👍

I'm a bit hesitant about the performance side of things as the crossfade issue sounds a bit like there could be more event firing happening than is intended. Given that there's been quite a few performance improvements in the editor landing of late, it might be worth double-checking with @ellatrix and @youknowriad to see if the `dragover` event listener sounds okay to go with for now. Since there can be multiple `BlockDraggable` components rendered on the page at a time, I'm wondering if it's possible that we're seeing competing callbacks firing, even though each individual callback is throttled?

Other than that performance issue, this is looking great to me! 🎉

tellthemachines commented:
> I do wish the performance of the crossfade was better, it feels sluggish to me

Could you share a visual of this? I'm finding it really hard to reproduce even with CPU throttling. Is it just about the CSS transition or the general movement of the chip?

andrewserong reviewed (commented):
Not sure if this is the same as Andy was seeing, but it is a little stuttery for me. A bit hard to show in a screengrab, but here's one I captured with CPU throttling (site editor with the list view opened):

https://github.com/WordPress/gutenberg/assets/14988353/f52c274f-3814-4837-96ac-0e0a9c6aa8aa

When hovering over the paragraph block, the crossed out icon appears to flicker on for a moment before going away.

In terms of the movement of the chip, anecdotally, after re-testing `trunk` it seems that the drag chip performance when dragging over the editor canvas feels much the same to me between this PR and `trunk` with CPU throttling enabled.

tellthemachines commented:
Ooooh ok, that stuttering isn't a performance issue, it's because any block that's not a container currently counts as an area that's not droppable in. I previously tried making sure the droppable areas were containers as a workaround for this, but that didn't work for root level areas, because drop indicators at root level seem to be attached to their sibling block client ids, as opposed to their containers.

I couldn't find a reliable workaround for this. Open to ideas!

jasmussen commented:
Nice improvements!

apeatling commented:
> Could you share a visual of this? I'm finding it really hard to reproduce even with CPU throttling. Is it just about the CSS transition or the general movement of the chip?

Not the movement of the chip, it's the crossfade between the two states within the chip:

https://github.com/WordPress/gutenberg/assets/1464705/735df6e6-0d67-4165-9f13-788fdf0c4850

andrewserong commented:
> I couldn't find a reliable workaround for this. Open to ideas!

> Not the movement of the chip, it's the crossfade between the two states within the chip:

What a tricky bit of detail! I haven't had much luck digging further into either of these so far, but something I observed with the stuttering is that it looks like the classnames flicker on and off a couple of times when hovering over a block. So that might be one part of the stuttering?

https://github.com/WordPress/gutenberg/assets/14988353/48179125-9790-4151-9d8f-f102ec5a0333

andrewserong commented:
Update: I have a hunch that I think might fix the stuttering, but in the process further exposes the issue of the disabled status being flagged when dragging over individual blocks. Here's what I could get running locally:

https://github.com/WordPress/gutenberg/assets/14988353/164ecd94-e285-40bf-862a-c6250a87c067

As far as I can tell, it seems that the use of the `useEffect` is resulting in `onDragOver` being fired too frequently, which I think is the cause of the stuttering / the classnames flickering on and off. One way to work around this could be to remove the `useEffect` and instead add a couple of small functions for attaching and removing the event listeners:

```
	const startDragOverBehavior = () => {
		editorRoot?.addEventListener( 'dragover', throttledOnDragOver );
	};

	const endDropOverBehavior = () => {
		editorRoot?.removeEventListener( 'dragover', throttledOnDragOver );
	};
```

Then, further down in the component we'd call these in the `Draggable` component's `onDragStart` and `onDragEnd` callbacks respectively. That way the event listener is only registered once, and only for the block being dragged.

In order to get `onDragOver` working without the `useEffect`, I then needed to construct `draggedBlockNames` within the `onDragOver` callback instead of outside it, so that we're not dependent on values that change, and can have a stable function for the throttle behaviour.

Here's a diff of where I got up to:

<details>

<summary>Diff of hacking around removing the `useEffect`</summary>

```diff
diff --git a/packages/block-editor/src/components/block-draggable/index.js b/packages/block-editor/src/components/block-draggable/index.js
index e5945841d1e..d509337d439 100644
--- a/packages/block-editor/src/components/block-draggable/index.js
+++ b/packages/block-editor/src/components/block-draggable/index.js
@@ -31,8 +31,9 @@ const BlockDraggable = ( {
 		visibleInserter,
 		getBlockType,
 		getAllowedBlocks,
-		draggedBlockNames,
+		// draggedBlockNames,
 		getBlockNamesByClientId,
+		getDraggedBlockClientIds,
 	} = useSelect(
 		( select ) => {
 			const {
@@ -41,7 +42,7 @@ const BlockDraggable = ( {
 				getBlockName,
 				getBlockAttributes,
 				isBlockInsertionPointVisible,
-				getDraggedBlockClientIds,
+				getDraggedBlockClientIds: _getDraggedBlockClientIds,
 				getBlockNamesByClientId: _getBlockNamesByClientId,
 				getAllowedBlocks: _getAllowedBlocks,
 			} = select( blockEditorStore );
@@ -62,9 +63,10 @@ const BlockDraggable = ( {
 				getBlockType: _getBlockType,
 				getAllowedBlocks: _getAllowedBlocks,
 				draggedBlockNames: _getBlockNamesByClientId(
-					getDraggedBlockClientIds()
+					_getDraggedBlockClientIds()
 				),
 				getBlockNamesByClientId: _getBlockNamesByClientId,
+				getDraggedBlockClientIds: _getDraggedBlockClientIds,
 			};
 		},
 		[ clientIds ]
@@ -90,63 +92,53 @@ const BlockDraggable = ( {
 	const blockRef = useBlockRef( clientIds[ 0 ] );
 	const editorRoot = blockRef.current?.closest( 'body' );

-	// Add a dragover event listener to the editor root to track the blocks being dragged over.
-	// The listener has to be inside the editor iframe otherwise the target isn't accessible.
-	// Check if the dragged blocks are allowed inside the target. If not, grey out the draggable.
-	useEffect( () => {
-		if ( ! editorRoot || ! fadeWhenDisabled ) {
+	const onDragOver = ( event ) => {
+		if ( ! event.target.closest( '[data-block]' ) ) {
 			return;
 		}

-		const onDragOver = ( event ) => {
-			if ( ! event.target.closest( '[data-block]' ) ) {
-				return;
-			}
-
-			const targetClientId = event.target
-				.closest( '[data-block]' )
-				.getAttribute( 'data-block' );
-
-			const allowedBlocks = getAllowedBlocks( targetClientId );
-			const targetBlockName = getBlockNamesByClientId( [
-				targetClientId,
-			] )[ 0 ];
-			const dropTargetValid = isDropTargetValid(
-				getBlockType,
-				allowedBlocks,
-				draggedBlockNames,
-				targetBlockName
+		const targetClientId = event.target
+			.closest( '[data-block]' )
+			.getAttribute( 'data-block' );
+
+		const draggedBlockNames = getBlockNamesByClientId(
+			getDraggedBlockClientIds()
+		);
+
+		const allowedBlocks = getAllowedBlocks( targetClientId );
+		const targetBlockName = getBlockNamesByClientId( [
+			targetClientId,
+		] )[ 0 ];
+		const dropTargetValid = isDropTargetValid(
+			getBlockType,
+			allowedBlocks,
+			draggedBlockNames,
+			targetBlockName
+		);
+
+		// Update the body class to reflect if drop target is valid.
+		// This has to be done on the document body because the draggable
+		// chip is rendered outside of the editor iframe.
+		if ( dropTargetValid && ! visibleInserter ) {
+			window?.document?.body?.classList?.add(
+				'block-draggable-invalid-drag-token'
 			);
+		} else {
+			window?.document?.body?.classList?.remove(
+				'block-draggable-invalid-drag-token'
+			);
+		}
+	};

-			// Update the body class to reflect if drop target is valid.
-			// This has to be done on the document body because the draggable
-			// chip is rendered outside of the editor iframe.
-			if ( ! dropTargetValid && ! visibleInserter ) {
-				window?.document?.body?.classList?.add(
-					'block-draggable-invalid-drag-token'
-				);
-			} else {
-				window?.document?.body?.classList?.remove(
-					'block-draggable-invalid-drag-token'
-				);
-			}
-		};
-
-		const throttledOnDragOver = throttle( onDragOver, 200 );
+	const throttledOnDragOver = throttle( onDragOver, 200 );

-		editorRoot.addEventListener( 'dragover', throttledOnDragOver );
+	const startDragOverBehavior = () => {
+		editorRoot?.addEventListener( 'dragover', throttledOnDragOver );
+	};

-		return () => {
-			editorRoot.removeEventListener( 'dragover', throttledOnDragOver );
-		};
-	}, [
-		draggedBlockNames,
-		editorRoot,
-		getAllowedBlocks,
-		getBlockNamesByClientId,
-		getBlockType,
-		visibleInserter,
-	] );
+	const endDropOverBehavior = () => {
+		editorRoot?.removeEventListener( 'dragover', throttledOnDragOver );
+	};

 	if ( ! isDraggable ) {
 		return children( { draggable: false } );
@@ -176,6 +168,7 @@ const BlockDraggable = ( {
 						onDragStart();
 					}
 				} );
+				startDragOverBehavior();
 			} }
 			onDragOver={ scrollOnDragOver }
 			onDragEnd={ () => {
@@ -187,6 +180,7 @@ const BlockDraggable = ( {
 				if ( onDragEnd ) {
 					onDragEnd();
 				}
+				endDropOverBehavior();
 			} }
 			__experimentalDragComponent={
 				<BlockDraggableChip
```

</details>

It's pretty messy and makes things a little more broken in the process, but just thought I'd share it in case it helps the investigating! Happy to keep looking tomorrow, too.

tellthemachines committed:
Only check validity of container blocks.

tellthemachines commented:
Thanks for all the testing and screengrabs folks! And thanks @andrewserong for diving in and looking at options ❤️ 

I think I've fixed it now. The problem wasn't as much in the effect itself - because it can run repeatedly without causing the body classname to get added or removed - but in the conditions we were checking. Because non-container blocks were flagged as non-droppable, dragging over them (particularly the Paragraphs for some reason) was causing the classname to get added and removed more than it should. I changed the check to look at the parent block when the current block isn't a container, and that seems to have done the trick!

I can't reproduce the crossfade issue @apeatling noticed anymore, so I think it might have been related? But let me know if you still see it!

andrewserong reviewed (approved):
Nicely done, that's resolved the issue for me!

> The problem wasn't as much in the effect itself - because it can run repeatedly without causing the body classname to get added or removed - but in the conditions we were checking

That seems consistent with what I'm seeing in testing now. The effect is still running more frequently than it needs to (i.e. because of the useEffect resulting in multiple handlers being called, it runs at greater than once every `200ms`), however every time it runs, the output is consistent, which means there's no longer a flicker of the classname being added and removed 👍

To check how frequently it's being called, I added a `let numTimesCalled = 0;` definition at the root of the file and called `console.log( ++numTimesCalled );` from within `onDragOver`. It's more obvious while dragging, but it gets called a fair bit:

https://github.com/WordPress/gutenberg/assets/14988353/d79f3675-4eec-4509-bf80-aa8028ec5aec

That doesn't seem like a blocker to me, as it seems to already be being called less frequently than once per frame, and removing the `useEffect` could be a good candidate for a code quality follow-up, rather than something that needs to block the feature getting in.

LGTM, great work with all the follow-ups here and figuring out all this nuanced logic! ✨

tellthemachines commented:
The only way I can think of to remove the effect is if we can decouple the callback from `visibleInserter`. I guess one thing we could look at is, instead of checking if the inserter is visible before adding the body class, adding _another_ body class when the inserter is visible (that could be added elsewhere in the actual inserter logic). The downside of that is we'd be adding that body class when it's not needed as it would be much more frequent than the invalid drop one. Not sure if that would be much of an optimisation 😅

andrewserong commented:
Yes, that does sound like a bit of a rabbithole (and a good argument for leaving it for now!). I'm happy to have a hack around at it in the new year if you'd like, but no need to delay this PR any longer for now IMO 🙂

tellthemachines committed:
Merge remote-tracking branch 'origin/trunk' into fix/drop-indicator-not-allowed

tellthemachines commented:
I just merged trunk into this branch as I think some checks were failing because of the node update.
